File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSameTree(TreeNode* first, TreeNode * second){
        if(first == nullptr && second == nullptr) return true;
        if(first == nullptr || second == nullptr) return false;
        if(first->val != second->val) return false;
        return isSameTree(first->left, second->left) && isSameTree(first->right, second->right);
    }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if (root == nullptr) return false;
        if(isSameTree(root,subRoot)) return true;
        return isSubtree(root->left,subRoot) || isSubtree(root->right, subRoot);
    }
};

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder :
            return None
        root= preorder[0]
        rootnode= TreeNode(root)
        rootidx= inorder.index(root)

        inorder_left= inorder[: rootidx]
        inorder_right= inorder[rootidx+1:]

        leftsize= len(inorder_left)
        preorder_left= preorder[1: 1+leftsize]
        preorder_right = preorder[1+ leftsize:]
        rootnode.left= self.buildTree(preorder_left, inorder_left)
        rootnode.right= self.buildTree(preorder_right, inorder_right)
        return rootnode